File: public/world.js
const NetworkUpdateRate = 50

const WorldPositiveX = 0
const WorldPositiveY = 1
const WorldPositiveZ = 2
const WorldNegativeX = 3
const WorldNegativeY = 4
const WorldNegativeZ = 5

class World {
    constructor(g, gl) {
        this.g = g
        this.gl = gl
        this.width
        this.height
        this.length
        this.slice
        this.all
        this.blocks
        this.viewable
        this.spriteSet
        this.spriteBuffer
        this.spriteCount
        this.thingCount
        this.itemCount
        this.missileCount
        this.particleCount
        this.things
        this.items
        this.missiles
        this.particles
        this.thingLookup
        this.itemLookup
        this.missileLookup
        this.occluder = new Occluder()
        this.PID
    }
    Load(raw) {
        this.blocks = []
        this.viewable = []
        this.spriteSet = new Set()
        this.spriteBuffer = {}
        this.spriteCount = {}
        this.thingCount = 0
        this.itemCount = 0
        this.missileCount = 0
        this.particleCount = 0
        this.things = []
        this.items = []
        this.missiles = []
        this.particles = []
        this.thingLookup = {}
        this.itemLookup = {}
        this.missileLookup = {}

        let dat = new DataView(raw)
        let dex = 0

        this.PID = dat.getUint16(dex, true)
        dex += 2

        this.width = dat.getUint16(dex, true)
        dex += 2
        this.height = dat.getUint16(dex, true)
        dex += 2
        this.length = dat.getUint16(dex, true)
        dex += 2
        this.slice = this.width * this.height
        this.all = this.slice * this.length

        let bx = 0
        let by = 0
        let bz = 0
        for (let b = 0; b < this.all; b++) {
            this.blocks[b] = new Block(bx, by, bz)
            bx++
            if (bx === this.width) {
                bx = 0
                by++
                if (by === this.height) {
                    by = 0
                    bz++
                }
            }
        }

        for (let b = 0; b < this.all; b++) {
            let block = this.blocks[b]
            let notEmpty = dat.getUint8(dex, true)
            dex += 1

            if (notEmpty) {
                for (let t = 0; t < BlockAll; t++) {
                    let tileType = dat.getUint8(dex, true)
                    dex += 1
                    block.tiles[t].type = tileType
                }
            }

            let thingCount = dat.getUint8(dex, true)
            dex += 1
            for (let t = 0; t < thingCount; t++) {
                let uid = dat.getUint16(dex, true)
                dex += 2
                switch (uid) {
                    case HumanUID:
                        {
                            let nid = dat.getUint16(dex, true)
                            dex += 2
                            let x = dat.getFloat32(dex, true)
                            dex += 4
                            let y = dat.getFloat32(dex, true)
                            dex += 4
                            let z = dat.getFloat32(dex, true)
                            dex += 4
                            let angle = dat.getFloat32(dex, true)
                            dex += 4
                            let health = dat.getUint16(dex, true)
                            dex += 2
                            if (nid === this.PID) new PlayerYou(this, nid, x, y, z, angle, health)
                            else new You(this, nid, x, y, z, angle, health)
                        }
                        continue
                    case BaronUID:
                        {
                            let nid = dat.getUint16(dex, true)
                            dex += 2
                            let x = dat.getFloat32(dex, true)
                            dex += 4
                            let y = dat.getFloat32(dex, true)
                            dex += 4
                            let z = dat.getFloat32(dex, true)
                            dex += 4
                            let direction = dat.getUint8(dex, true)
                            dex += 1
                            let health = dat.getUint16(dex, true)
                            dex += 2
                            let status = dat.getUint8(dex, true)
                            dex += 1
                            new Baron(this, nid, x, y, z, direction, health, status)
                        }
                        continue
                    case TreeUID:
                        {
                            let nid = dat.getUint16(dex, true)
                            dex += 2
                            let x = dat.getFloat32(dex, true)
                            dex += 4
                            let y = dat.getFloat32(dex, true)
                            dex += 4
                            let z = dat.getFloat32(dex, true)
                            dex += 4
                            new Tree(this, nid, x, y, z)
                        }
                        continue
                }
            }

            let itemCount = dat.getUint8(dex, true)
            for (let t = 0; t < itemCount; t++) {
                console.log("new item")
            }
            dex += 1

            let missileCount = dat.getUint8(dex, true)
            dex += 1
            for (let t = 0; t < missileCount; t++) {
                let uid = dat.getUint16(dex, true)
                dex += 2
                switch (uid) {
                    case PlasmaUID:
                        {
                            let nid = dat.getUint16(dex, true)
                            dex += 2
                            let x = dat.getFloat32(dex, true)
                            dex += 4
                            let y = dat.getFloat32(dex, true)
                            dex += 4
                            let z = dat.getFloat32(dex, true)
                            dex += 4
                            let dx = dat.getFloat32(dex, true)
                            dex += 4
                            let dy = dat.getFloat32(dex, true)
                            dex += 4
                            let dz = dat.getFloat32(dex, true)
                            dex += 4
                            let damage = dat.getUint16(dex, true)
                            dex += 2
                            new Plasma(this, nid, damage, x, y, z, dx, dy, dz)
                        }
                        continue
                }
            }

            let lightCount = dat.getUint8(dex, true)
            dex += 1
            for (let t = 0; t < lightCount; t++) {
                let x = dat.getUint8(dex, true)
                dex += 1
                let y = dat.getUint8(dex, true)
                dex += 1
                let z = dat.getUint8(dex, true)
                dex += 1
                let rgb = dat.getInt32(dex, true)
                dex += 4
                block.AddLight(new Light(x, y, z, rgb))
            }
        }

        this.build()
    }
    build() {
        for (let i = 0; i < this.all; i++) {
            let block = this.blocks[i]
            for (let j = 0; j < block.lights.length; j++)
                Light.Add(this, block, block.lights[j])
            Occluder.SetBlockVisible(block)
        }
        for (let i = 0; i < this.all; i++)
            this.blocks[i].BuildMesh(this)
    }
    FindBlock(x, y, z) {
        let gx = Math.floor(x)
        let gy = Math.floor(y)
        let gz = Math.floor(z)
        let bx = Math.floor(gx * InverseBlockSize)
        let by = Math.floor(gy * InverseBlockSize)
        let bz = Math.floor(gz * InverseBlockSize)
        let tx = gx - bx * BlockSize
        let ty = gy - by * BlockSize
        let tz = gz - bz * BlockSize
        let block = this.blocks[bx + by * this.width + bz * this.slice]
        return block.tiles[tx + ty * BlockSize + tz * BLOCK_SLICE].type
    }
    GetTilePointer(cx, cy, cz, bx, by, bz) {
        while (bx < 0) {
            bx += BlockSize
            cx--
        }
        while (bx >= BlockSize) {
            bx -= BlockSize
            cx++
        }
        while (by < 0) {
            by += BlockSize
            cy--
        }
        while (by >= BlockSize) {
            by -= BlockSize
            cy++
        }
        while (bz < 0) {
            bz += BlockSize
            cz--
        }
        while (bz >= BlockSize) {
            bz -= BlockSize
            cz++
        }
        let block = this.GetBlock(cx, cy, cz)
        if (block === null)
            return null
        return block.GetTilePointerUnsafe(bx, by, bz)
    }
    GetTileType(cx, cy, cz, bx, by, bz) {
        while (bx < 0) {
            bx += BlockSize
            cx--
        }
        while (bx >= BlockSize) {
            bx -= BlockSize
            cx++
        }
        while (by < 0) {
            by += BlockSize
            cy--
        }
        while (by >= BlockSize) {
            by -= BlockSize
            cy++
        }
        while (bz < 0) {
            bz += BlockSize
            cz--
        }
        while (bz >= BlockSize) {
            bz -= BlockSize
            cz++
        }
        let block = this.GetBlock(cx, cy, cz)
        if (block === null) {
            return TILE_NONE
        }
        return block.GetTileTypeUnsafe(bx, by, bz)
    }
    GetBlock(x, y, z) {
        if (x < 0 || x >= this.width)
            return null
        if (y < 0 || y >= this.height)
            return null
        if (z < 0 || z >= this.length)
            return null
        return this.blocks[x + y * this.width + z * this.slice]
    }
    AddThing(thing) {
        this.things[this.thingCount] = thing
        this.thingCount++
        this.thingLookup[thing.NID] = thing

        let count = this.spriteCount[thing.SID]
        if (count) {
            this.spriteCount[thing.SID] = count + 1
            if ((count + 2) * 16 > this.spriteBuffer[thing.SID].vertices.length)
                this.spriteBuffer[thing.SID] = RenderBuffer.Expand(this.gl, this.spriteBuffer[thing.SID])
        } else {
            this.spriteCount[thing.SID] = 1
            this.spriteBuffer[thing.SID] = RenderBuffer.Init(this.gl, 3, 0, 2, 40, 60)
        }
    }
    AddItem(item) {
        this.items[this.itemCount] = item
        this.itemCount++
        this.itemLookup[item.NID] = item

        let count = this.spriteCount[item.SID]
        if (count) {
            this.spriteCount[item.SID] = count + 1
            if ((count + 2) * 16 > this.spriteBuffer[item.SID].vertices.length)
                this.spriteBuffer[item.SID] = RenderBuffer.Expand(this.gl, this.spriteBuffer[item.SID])
        } else {
            this.spriteCount[item.SID] = 1
            this.spriteBuffer[item.SID] = RenderBuffer.Init(this.gl, 3, 0, 2, 40, 60)
        }
    }
    AddMissile(missile) {
        this.missiles[this.missileCount] = missile
        this.missileCount++
        this.missileLookup[missile.NID] = missile

        let count = this.spriteCount[missile.SID]
        if (count) {
            this.spriteCount[missile.SID] = count + 1
            if ((count + 2) * 16 > this.spriteBuffer[missile.SID].vertices.length)
                this.spriteBuffer[missile.SID] = RenderBuffer.Expand(this.gl, this.spriteBuffer[missile.SID])
        } else {
            this.spriteCount[missile.SID] = 1
            this.spriteBuffer[missile.SID] = RenderBuffer.Init(this.gl, 3, 0, 2, 40, 60)
        }
    }
    AddParticle(particle) {
        this.particles[this.particleCount] = particle
        this.particleCount++

        let count = this.spriteCount[particle.SID]
        if (count) {
            this.spriteCount[particle.SID] = count + 1
            if ((count + 2) * 16 > this.spriteBuffer[particle.SID].vertices.length) {
                this.spriteBuffer[particle.SID] = RenderBuffer.Expand(this.gl, this.spriteBuffer[particle.SID])
            }
        } else {
            this.spriteCount[particle.SID] = 1
            this.spriteBuffer[particle.SID] = RenderBuffer.Init(this.gl, 3, 0, 2, 120, 180)
        }
    }
    RemoveThing(thing) {
        let len = this.thingCount
        for (let i = 0; i < len; i++) {
            if (this.things[i] === thing) {
                this.things[i] = this.things[len - 1]
                this.things[len - 1] = null
                this.thingCount--
                this.spriteCount[thing.SID]--
                delete this.thingLookup[thing.NID]
                break
            }
        }
    }
    RemoveItem(item) {
        let len = this.itemCount
        for (let i = 0; i < len; i++) {
            if (this.items[i] === item) {
                this.items[i] = this.items[len - 1]
                this.items[len - 1] = null
                this.itemCount--
                this.spriteCount[item.SID]--
                delete this.itemLookup[item.NID]
                break
            }
        }
    }
    RemoveMissile(missile) {
        let len = this.missileCount
        for (let i = 0; i < len; i++) {
            if (this.missiles[i] === missile) {
                this.missiles[i] = this.missiles[len - 1]
                this.missiles[len - 1] = null
                this.missileCount--
                this.spriteCount[missile.SID]--
                delete this.missileLookup[missile.NID]
                break
            }
        }
    }
    RemoveParticle(particle) {
        let len = this.particleCount
        for (let i = 0; i < len; i++) {
            if (this.particles[i] === particle) {
                this.particles[i] = this.particles[len - 1]
                this.particles[len - 1] = null
                this.particleCount--
                this.spriteCount[particle.SID]--
                break
            }
        }
    }
    update() {
        let len = this.thingCount
        for (let i = 0; i < len; i++)
            this.things[i].Update()
        len = this.missileCount
        for (let i = 0; i < len; i++) {
            if (this.missiles[i].Update()) {
                this.missiles[i] = this.missiles[len - 1]
                this.missiles[len - 1] = null
                this.missileCount--
                len--
                i--
            }
        }
        len = this.particleCount
        for (let i = 0; i < this.particleCount; i++) {
            if (this.particles[i].Update()) {
                this.particles[i] = this.particles[len - 1]
                this.particles[len - 1] = null
                this.particleCount--
                len--
                i--

            }
        }
    }
    render(g, interpolation, x, y, z, camX, camZ, camAngle) {
        let gl = this.gl
        let spriteSet = this.spriteSet
        let spriteBuffer = this.spriteBuffer

        this.occluder.prepare_frustum(g)
        this.occluder.occlude(this, x, y, z)

        spriteSet.clear()
        for (let key in spriteBuffer)
            spriteBuffer[key].Zero()

        g.set_program(gl, "texcol3d")
        g.update_mvp(gl)
        g.set_texture(gl, "tiles")

        for (let i = 0; i < OCCLUSION_VIEW_NUM; i++) {
            let block = this.viewable[i]

            block.RenderThings(interpolation, spriteSet, spriteBuffer, camX, camZ, camAngle)

            let mesh = block.mesh
            if (mesh.vertex_pos === 0)
                continue

            RenderSystem.BindVao(gl, mesh)

            if (x == block.x) {
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveX], block.count_side[WorldPositiveX])
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeX], block.count_side[WorldNegativeX])
            } else if (x > block.x)
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveX], block.count_side[WorldPositiveX])
            else
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeX], block.count_side[WorldNegativeX])

            if (y == block.y) {
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveY], block.count_side[WorldPositiveY])
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeY], block.count_side[WorldNegativeY])
            } else if (y > block.y)
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveY], block.count_side[WorldPositiveY])
            else
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeY], block.count_side[WorldNegativeY])

            if (z == block.z) {
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveZ], block.count_side[WorldPositiveZ])
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeZ], block.count_side[WorldNegativeZ])
            } else if (z > block.z)
                RenderSystem.DrawRange(gl, block.begin_side[WorldPositiveZ], block.count_side[WorldPositiveZ])
            else
                RenderSystem.DrawRange(gl, block.begin_side[WorldNegativeZ], block.count_side[WorldNegativeZ])
        }

        g.set_program(gl, "texture3d")
        g.update_mvp(gl)
        for (let key in spriteBuffer) {
            let buffer = spriteBuffer[key]
            if (buffer.vertex_pos > 0) {
                g.set_texture(gl, key)
                RenderSystem.UpdateAndDraw(gl, buffer)
            }
        }
    }
}

File: client/world.go
package main

import (
	"bytes"
	"encoding/binary"
	"syscall/js"

	"./graphics"
)

// Constants
const (
	NetUpdateRate  = 50
	InverseNetRate = 16.67 / NetUpdateRate

	WorldPositiveX = 0
	WorldPositiveY = 1
	WorldPositiveZ = 2
	WorldNegativeX = 3
	WorldNegativeY = 4
	WorldNegativeZ = 5

	BroadcastNew    = 0
	BroadcastDelete = 1
	BroadcastChat   = 2
)

type world struct {
	g             *graphics.RenderSystem
	gl            js.Value
	width         int
	height        int
	length        int
	tileWidth     int
	tileHeight    int
	tileLength    int
	slice         int
	all           int
	blocks        []block
	occluder      *occluder
	spriteBuffer  map[string]*graphics.RenderBuffer
	spriteCount   map[string]int
	thingCount    int
	itemCount     int
	missileCount  int
	particleCount int
	things        []*thing
	items         []*item
	missiles      []*missile
	particles     []*particle
	netLookup     map[uint16]interface{}
	pid           uint16
}

func worldInit(g *graphics.RenderSystem, gl js.Value) *world {
	world := &world{}
	world.g = g
	world.gl = gl
	world.spriteBuffer = make(map[string]*graphics.RenderBuffer)
	return world
}

func (me *world) reset() {
	me.blocks = make([]block, me.all)
	me.spriteCount = make(map[string]int)
	me.thingCount = 0
	me.itemCount = 0
	me.missileCount = 0
	me.particleCount = 0
	me.things = make([]*thing, 0)
	me.items = make([]*item, 0)
	me.missiles = make([]*missile, 0)
	me.particles = make([]*particle, 0)
	me.netLookup = make(map[uint16]interface{})
}

func (me *world) load(raw []byte) {
	dat := bytes.NewReader(raw)

	var uint8ref uint8
	var uint16ref uint16

	// TODO binary.Read sucks, implement custom solution for client and server
	// https://github.com/golang/go/blob/master/src/encoding/binary/binary.go

	binary.Read(dat, binary.LittleEndian, &me.pid)

	binary.Read(dat, binary.LittleEndian, &uint16ref)
	me.width = int(uint16ref)
	binary.Read(dat, binary.LittleEndian, &uint16ref)
	me.height = int(uint16ref)
	binary.Read(dat, binary.LittleEndian, &uint16ref)
	me.length = int(uint16ref)

	me.slice = me.width * me.height
	me.all = me.slice * me.length

	me.tileWidth = me.width * BlockSize
	me.tileHeight = me.height * BlockSize
	me.tileLength = me.length * BlockSize

	me.reset()
	me.occluder = occluderInit(me.all)

	bx := 0
	by := 0
	bz := 0
	for i := 0; i < me.all; i++ {
		me.blocks[i].blockInit(bx, by, bz)
		bx++
		if bx == me.width {
			bx = 0
			by++
			if by == me.height {
				by = 0
				bz++
			}
		}
	}

	for i := 0; i < me.all; i++ {
		block := &me.blocks[i]
		binary.Read(dat, binary.LittleEndian, &uint8ref)
		notEmpty := uint8ref != 0
		if notEmpty {
			for t := 0; t < BlockAll; t++ {
				binary.Read(dat, binary.LittleEndian, &uint8ref)
				tileType := int(uint8ref)
				block.tiles[t].typeOf = tileType
			}
		}

		var lightCount uint8
		binary.Read(dat, binary.LittleEndian, &lightCount)
		for t := uint8(0); t < lightCount; t++ {
			var x uint8
			var y uint8
			var z uint8
			var rgb int32
			binary.Read(dat, binary.LittleEndian, &x)
			binary.Read(dat, binary.LittleEndian, &y)
			binary.Read(dat, binary.LittleEndian, &z)
			binary.Read(dat, binary.LittleEndian, &rgb)
			block.addLight(lightInit(x, y, z, rgb))
		}
	}

	var thingCount uint16
	binary.Read(dat, binary.LittleEndian, &thingCount)
	for t := uint16(0); t < thingCount; t++ {
		var uid uint16
		var nid uint16
		var x float32
		var y float32
		var z float32
		binary.Read(dat, binary.LittleEndian, &uid)
		binary.Read(dat, binary.LittleEndian, &nid)
		binary.Read(dat, binary.LittleEndian, &x)
		binary.Read(dat, binary.LittleEndian, &y)
		binary.Read(dat, binary.LittleEndian, &z)
		switch uid {
		case HumanUID:
			var angle float32
			var health uint16
			var status uint8
			binary.Read(dat, binary.LittleEndian, &angle)
			binary.Read(dat, binary.LittleEndian, &health)
			binary.Read(dat, binary.LittleEndian, &status)
			if nid == me.pid {
				youInit(me, nid, x, y, z, angle, health, status)
			} else {
				humanInit(me, nid, x, y, z, angle, health, status)
			}
		case BaronUID:
			var direction uint8
			var health uint16
			var status uint8
			binary.Read(dat, binary.LittleEndian, &direction)
			binary.Read(dat, binary.LittleEndian, &health)
			binary.Read(dat, binary.LittleEndian, &status)
			baronInit(me, nid, x, y, z, direction, health, status)
		case TreeUID:
			treeInit(me, nid, x, y, z)
		}
	}

	var itemCount uint16
	binary.Read(dat, binary.LittleEndian, &itemCount)
	for t := uint16(0); t < itemCount; t++ {
		var uid uint16
		var nid uint16
		var x float32
		var y float32
		var z float32
		binary.Read(dat, binary.LittleEndian, &uid)
		binary.Read(dat, binary.LittleEndian, &nid)
		binary.Read(dat, binary.LittleEndian, &x)
		binary.Read(dat, binary.LittleEndian, &y)
		binary.Read(dat, binary.LittleEndian, &z)
		switch uid {
		case MedkitUID:
			medkitInit(me, nid, x, y, z)
		}
	}

	var missileCount uint16
	binary.Read(dat, binary.LittleEndian, &missileCount)
	for t := uint16(0); t < missileCount; t++ {
		var uid uint16
		var nid uint16
		var x float32
		var y float32
		var z float32
		var dx float32
		var dy float32
		var dz float32
		binary.Read(dat, binary.LittleEndian, &uid)
		binary.Read(dat, binary.LittleEndian, &nid)
		binary.Read(dat, binary.LittleEndian, &x)
		binary.Read(dat, binary.LittleEndian, &y)
		binary.Read(dat, binary.LittleEndian, &z)
		binary.Read(dat, binary.LittleEndian, &dx)
		binary.Read(dat, binary.LittleEndian, &dy)
		binary.Read(dat, binary.LittleEndian, &dz)
		switch uid {
		case PlasmaUID:
			var damage uint16
			binary.Read(dat, binary.LittleEndian, &damage)
			plasmaInit(me, nid, damage, x, y, z, dx, dy, dz)
		}
	}

	me.build()
}

func (me *world) build() {
	for i := 0; i < me.all; i++ {
		block := &me.blocks[i]
		for j := 0; j < block.lightCount; j++ {
			block.lights[j].addToWorld(me, block)
		}
		block.occlusion()
	}
	for i := 0; i < me.all; i++ {
		me.blocks[i].buildMesh(me)
	}
}

func (me *world) getTilePointer(bx, by, bz, tx, ty, tz int) *tile {
	for tx < 0 {
		tx += BlockSize
		bx--
	}
	for tx >= BlockSize {
		tx -= BlockSize
		bx++
	}
	for ty < 0 {
		ty += BlockSize
		by--
	}
	for ty >= BlockSize {
		ty -= BlockSize
		by++
	}
	for tz < 0 {
		tz += BlockSize
		bz--
	}
	for tz >= BlockSize {
		tz -= BlockSize
		bz++
	}
	block := me.getBlock(bx, by, bz)
	if block == nil {
		return nil
	}
	return &block.tiles[tx+ty*BlockSize+tz*BlockSlice]
}

func (me *world) getTileType(bx, by, bz, tx, ty, tz int) int {
	for tx < 0 {
		tx += BlockSize
		bx--
	}
	for tx >= BlockSize {
		tx -= BlockSize
		bx++
	}
	for ty < 0 {
		ty += BlockSize
		by--
	}
	for ty >= BlockSize {
		ty -= BlockSize
		by++
	}
	for tz < 0 {
		tz += BlockSize
		bz--
	}
	for tz >= BlockSize {
		tz -= BlockSize
		bz++
	}
	block := me.getBlock(bx, by, bz)
	if block == nil {
		return TileNone
	}
	return block.tiles[tx+ty*BlockSize+tz*BlockSlice].typeOf
}

func (me *world) getBlock(x, y, z int) *block {
	if x < 0 || x >= me.width {
		return nil
	}
	if y < 0 || y >= me.height {
		return nil
	}
	if z < 0 || z >= me.length {
		return nil
	}
	return &me.blocks[x+y*me.width+z*me.slice]
}

func (me *world) addThing(t *thing) {
	if me.thingCount == len(me.things) {
		array := make([]*thing, me.thingCount+5)
		copy(array, me.things)
		me.things = array
	}
	me.things[me.thingCount] = t
	me.thingCount++

	count, has := me.spriteCount[t.sid]
	if has {
		me.spriteCount[t.sid] = count + 1
		b := me.spriteBuffer[t.sid]
		if (count+2)*16 > len(b.Vertices) {
			b.RenderBufferExpand(me.gl)
		}
	} else {
		me.spriteCount[t.sid] = 1
		me.spriteBuffer[t.sid] = graphics.RenderBufferInit(me.gl, 3, 0, 2, 40, 60)
	}
}

func (me *world) addItem(t *item) {
	if me.itemCount == len(me.items) {
		array := make([]*item, me.itemCount+5)
		copy(array, me.items)
		me.items = array
	}
	me.items[me.itemCount] = t
	me.itemCount++

	me.netLookup[t.nid] = t

	count, has := me.spriteCount[t.sid]
	if has {
		me.spriteCount[t.sid] = count + 1
		b := me.spriteBuffer[t.sid]
		if (count+2)*16 > len(b.Vertices) {
			b.RenderBufferExpand(me.gl)
		}
	} else {
		me.spriteCount[t.sid] = 1
		me.spriteBuffer[t.sid] = graphics.RenderBufferInit(me.gl, 3, 0, 2, 40, 60)
	}
}

func (me *world) addMissile(t *missile) {
	if me.missileCount == len(me.missiles) {
		array := make([]*missile, me.missileCount+5)
		copy(array, me.missiles)
		me.missiles = array
	}
	me.missiles[me.missileCount] = t
	me.missileCount++

	me.netLookup[t.nid] = t

	count, has := me.spriteCount[t.sid]
	if has {
		me.spriteCount[t.sid] = count + 1
		b := me.spriteBuffer[t.sid]
		if (count+2)*16 > len(b.Vertices) {
			b.RenderBufferExpand(me.gl)
		}
	} else {
		me.spriteCount[t.sid] = 1
		me.spriteBuffer[t.sid] = graphics.RenderBufferInit(me.gl, 3, 0, 2, 40, 60)
	}
}

func (me *world) addParticle(t *particle) {
	if me.particleCount == len(me.particles) {
		array := make([]*particle, me.particleCount+5)
		copy(array, me.particles)
		me.particles = array
	}
	me.particles[me.particleCount] = t
	me.particleCount++

	count, has := me.spriteCount[t.sid]
	if has {
		me.spriteCount[t.sid] = count + 1
		b := me.spriteBuffer[t.sid]
		if (count+2)*16 > len(b.Vertices) {
			b.RenderBufferExpand(me.gl)
		}
	} else {
		me.spriteCount[t.sid] = 1
		me.spriteBuffer[t.sid] = graphics.RenderBufferInit(me.gl, 3, 0, 2, 40, 60)
	}
}

func (me *world) removeThing(t *thing) {
	size := me.thingCount
	for i := 0; i < size; i++ {
		if me.things[i] == t {
			me.things[i] = me.things[size-1]
			me.things[size-1] = nil
			me.thingCount--
			break
		}
	}
}

func (me *world) removeItem(t *item) {
	size := me.itemCount
	for i := 0; i < size; i++ {
		if me.items[i] == t {
			me.items[i] = me.items[size-1]
			me.items[size-1] = nil
			me.itemCount--
			break
		}
	}
}

func (me *world) removeMissile(t *missile) {
	size := me.missileCount
	for i := 0; i < size; i++ {
		if me.missiles[i] == t {
			me.missiles[i] = me.missiles[size-1]
			me.missiles[size-1] = nil
			me.missileCount--
			break
		}
	}
}

func (me *world) removeParticle(t *particle) {
	size := me.particleCount
	for i := 0; i < size; i++ {
		if me.particles[i] == t {
			me.particles[i] = me.particles[size-1]
			me.particles[size-1] = nil
			me.particleCount--
			break
		}
	}
}

func (me *world) update() {
	size := me.thingCount
	for i := 0; i < size; i++ {
		me.things[i].update()
	}
	size = me.missileCount
	for i := 0; i < size; i++ {
		if me.missiles[i].update() {
			me.missiles[i] = me.missiles[size-1]
			me.missiles[size-1] = nil
			me.missileCount--
			size--
			i--
		}
	}
	size = me.particleCount
	for i := 0; i < size; i++ {
		if me.particles[i].update() {
			me.particles[i] = me.particles[size-1]
			me.particles[size-1] = nil
			me.particleCount--
			size--
			i--
		}
	}
}

func (me *world) render(g *graphics.RenderSystem, x, y, z int, camX, camZ, camAngle float32) {
	gl := me.gl
	spriteBuffer := me.spriteBuffer
	occluder := me.occluder
	spriteSet := make(map[interface{}]bool)

	occluder.prepareFrustum(g)
	occluder.search(me, x, y, z)

	for _, buffer := range spriteBuffer {
		buffer.Zero()
	}

	g.SetProgram(gl, "texture-color3d")
	g.UpdateMvp(gl)
	g.SetTexture(gl, "tiles")

	size := occluder.viewNum
	for i := 0; i < size; i++ {
		block := occluder.viewable[i]
		block.renderThings(spriteSet, spriteBuffer, camX, camZ, camAngle)
		mesh := block.mesh
		if mesh.VertexPos == 0 {
			continue
		}

		graphics.RenderSystemBindVao(gl, mesh)

		if x == block.x {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveX], block.countSide[WorldPositiveX])
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeX], block.countSide[WorldNegativeX])
		} else if x > block.x {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveX], block.countSide[WorldPositiveX])
		} else {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeX], block.countSide[WorldNegativeX])
		}

		if y == block.y {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveY], block.countSide[WorldPositiveY])
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeY], block.countSide[WorldNegativeY])
		} else if y > block.y {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveY], block.countSide[WorldPositiveY])
		} else {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeY], block.countSide[WorldNegativeY])
		}

		if z == block.z {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveZ], block.countSide[WorldPositiveZ])
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeZ], block.countSide[WorldNegativeZ])
		} else if z > block.z {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldPositiveZ], block.countSide[WorldPositiveZ])
		} else {
			graphics.RenderSystemDrawRange(gl, block.beginSide[WorldNegativeZ], block.countSide[WorldNegativeZ])
		}
	}

	g.SetProgram(gl, "texture3d")
	g.UpdateMvp(gl)
	for name, buffer := range spriteBuffer {
		if buffer.VertexPos > 0 {
			g.SetTexture(gl, name)
			graphics.RenderSystemUpdateAndDraw(gl, buffer)
		}
	}
}
